### DISCUSSION TITLE
Migrate to C/C++ for embedded systems

### DISCUSSION THREAD

**Discussion by GermanAizek** [vosen/ZLUDA#143] [Ideas]
Does anyone have plans to rewrite this in C++ or C for integration into embedded devices?

**Comment by cb88** [comment]
Why... that would be going backwards for no reason. It's written in rust that doesn't stop you from running it on any embedded device that can also run a GPU.

**Comment by GermanAizek** [comment]
> Why... that would be going backwards for no reason. It's written in rust that doesn't stop you from running it on any embedded device that can also run a GPU.

is it possible to run it on OpenWRT with MIPS32 24Kc architecture, for example?

https://forum.openwrt.org/t/installing-rust-compiler-on-openwrt/182937/5

in short:
```sh
$ ./rustup-init.sh
info: downloading installer
Warning: Not enforcing strong cipher suites for TLS, this is potentially less secure
curl: (22) The requested URL returned error: 404
rustup: installer for platform 'riscv64gc-unknown-linux-musl' not found, this may be unsupported
```

on doc.rust-lang.org i find toolchains, but I didn't find any for MIPS32 and RISCV64. Most likely they have not been written yet.

![image](https://github.com/vosen/ZLUDA/assets/21138600/e24f602c-c644-4ff2-8660-cde6d7bd2f5c)

**Comment by cb88** [comment]
What hardware are you talking about because MIPS32 is pretty much dead?

The RISCV64 system can work already just pick a distro that uses GLIBC instead of MUSL and call it a day. It's a tier 2 platform with essentially full support. 

In any case the real gottcha is probably not even rust but ROCm...

**Comment by nonetrix** [comment]
What embedded device has a modern AMD GPU? Just curious

**Comment by GermanAizek** [comment]
> What hardware are you talking about because MIPS32 is pretty much dead?

@cb88,
Since when has it stopped being supported, most switches, firewalls and routers are based on this architecture. You probably have only one live platform, it's `x86-64`, Rust really still can't compile for `MIPS32` and `RISCV64`, what you're offering is just not working snapshot toolchain.

**Comment by cb88** [comment]
@GermanAizek Again... Both MIP 32 LE and RISCV64 have supported toolchains. MIPS32 LE you have to build yourself just like any other embedded toolchain, and they provide prebuilt RISCV64 toolchains for a common toolchain version using GLIBC where ROCm will also work.... MUSL is not a common c library for this type of software. If you can't build your own embedded toolchains thats a you problem not something you should blame on others.

I guarantee you ROCm doesn't build on MUSL either in fact there are closed bug reports about it not working (closed without adding support).

And you still haven't referenced a realistic device where you would be connecting a GPU to a MIPS32 CPU... I'm certainly not aware of any.

**Comment by Kreijstal** [comment]
your embedded devices have GPUs?

**Reply by erkinalp** [reply]
videocore for raspberrypi, and arm mali for STBs and mobile phones

**Reply by Kreijstal** [reply]
are those embedded devices?

**Reply by erkinalp** [reply]
yes, raspberrypi is not, but other configurations using videocore and mali are

**Reply by Kreijstal** [reply]
does cuda work there?

**Reply by cb88** [reply]
Rust is available for those platforms....as are binary distros, presumably ZLUDA would just be another binary on those systems built with rust. That said it would only work if you added support for those GPUs or were able to wire up an AMD GPU (some of the recent boards do have a few lanes of PCIe) In any case it still isn't a reason to rewrite it in C++...since rust is already available there.

**Reply by nonetrix** [reply]
That sounds out of scope really, I mean if someone wanted to make a emulation layer for CUDA on the Rasberry Pi that would be rad actually. Not sure what the use case would be, but I guess I implore anyone to do so not sure what the GPU compute situation is like there though. But could a CUDA emulation layer be made with OpenCL or Vulkan or anything cross GPU? That would be super cool, again, not sure how OpenCL is on Rasberry Pi but Vulkan is definitely there but not sure about the compute part of Vulkan

Vulkan is the solution to the CUDA problem imo just some examples
https://github.com/DTolm/VkFFT
https://github.com/xinntao/Real-ESRGAN-ncnn-vulkan - I used this one a lot until the Nix package broke :/
I'd like to see more AI projects using Vulkan, I wish you could run LLMs or image generation that would be great

Edit found project for LLMs on Vulkan:
https://github.com/mlc-ai/mlc-llm

**Comment by vosen** [comment]
Converting to discussion because there's no bug that can be fixed here

**Reply by erkinalp** [reply]
no task that can be assigned out either?

**Reply by deuszex** [reply]
I'm pretty sure there is nothing to assign here. Open-source project, I'm pretty sure if someone wants to do a c/c++ implementation then they are free to do so based on license.

**Comment by codiggermkt** [comment]
Is anyone interested in CIDE?
Our integrated architecture harmonizes development, testing and production environments.
Codigger is a distributed operating system, enabling users from all over the world to collaborate for work, development and entertainment online. It also supports the creation and operation of web based personal applications.
